Proposal for Issue #12 - qos-booking-and-assignment.yaml

RamTMO opened the pull request:
#### What type of PR is this?

Add one of the following kinds:
* enhancement/feature



#### What this PR does / why we need it:
This is  proposal for QoS Booking and add a solution issue #12. It can live with QoS Booking to address complex use cases.



#### Which issue(s) this PR fixes:

<!-- Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`. -->

Fixes #12 

#### Special notes for reviewers:
Please review and I would really appriciate if you can pull this in for Fall'25.  I will prepare the next set of manditory deliverables as needed.


#### Changelog input

```
 Initial Version of qos-booking-and-assignment.yaml
```

#### Additional documentation 

This section can be blank.



```
docs

```

jlurien commented on code/API_definitions/qos-booking-and-assignment.yaml:
```diff
@@ -0,0 +1,1644 @@
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/BookingOutput"
```
For this operation we may just have numDevices instead of totalDevicesInBooking and remainingDevicesInBooking

RamTMO replied:
BookingOutput is used; and naturally has these two properties.  In this case, the remaining devices = total devices.

jlurien replied:
Acceptable to save an extra schema

jlurien commented on code/API_definitions/qos-booking-and-assignment.yaml:
```diff
@@ -0,0 +1,1644 @@
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/DeviceAssignmentOutput"
```
We have to think on how to return details on assigned devices and deal with the rules we have regarding use of 3-legged tokens. We may need to think if operations for QoS Booking (not device assignment) may sense for 3-legged and how to deal with that.

RamTMO replied:
The output will not disclose devices that are assigned via 3-legged tokens. I have added lots of text in the code. Please review.  Hopefully this will take care of this.

jlurien commented on code/API_definitions/qos-booking-and-assignment.yaml:
```diff
@@ -0,0 +1,1644 @@
+        - openId:
+            - "qos-booking-and-assignment:existing-qos-booking:delete"
+
+  /devices/assign:
```
maybe `/qos-bookings/{bookingId}/devices/assign` or even just `/qos-bookings/{bookingId}/devices`

RamTMO replied:
Please review and see if you are comfortable with.  We can discuss if needed.

jlurien commented on code/API_definitions/qos-booking-and-assignment.yaml:
```diff
@@ -0,0 +1,1644 @@
+        content:
+          application/json:
+            schema:
+              $ref: "#/components/schemas/DeviceAssignmentInput"
```
Here we have to think also how to deal with 3-legged tokens rules. If we allow 3-legged tokens then devices should not be sent, and only in case of 2-legged tokens we could allow several devices to be sent at once

I wouldn't allow to change sink and sinkCredentials with this operation, unless we want to have granularity for notiifcation at device level, which would be complex

RamTMO replied:
Your later comments answer to this question.  We can keep it seperate and mostly applicable for 2-legged scenarios.  We should review this in the context of 3-legged scenarios and make changes as needed in the next revision. 

I kept the sink as it is as an optional input. For example, I have a booking for 10 devices.  I ask to assign 4 devices; the operator needs to validate the devices before sending SUCCESSFUL message.  In this case, a PENDING will be returned and call back will come back with SUCCESSFUL, or PARTIAL_SUCCESS or FAILURE (or DECLINED).  The user's app can call GET anytime with the bookingID to findout the list of devices that are assigned to the booking.

jlurien commented on code/API_definitions/qos-booking-and-assignment.yaml:
```diff
@@ -0,0 +1,1644 @@
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/DeviceAssignmentOutput"
```
Regrading output, problem is when device is provided by means of a 3-legged token, in which case details (phoneNumber. IPs), should not be disclosed according to CAMARA guidelines, but this is problematic for the customer as they would miss information to know which devices are successfully attached.

I think that in general this API works better in 2-legged scenarios, while for 3-legged the other original API makes more sense.

RamTMO replied:
Agree. In that context, there is an endpoint /devices/retrieve I left as it is in my version to be backward compatible with yours.. Let me know if you want to keep it or can remove it

jlurien replied:
The question here is to document which device details must be returned here. For example, client assigns 3 devices to the booking, 2 in a call with 2-legged and 1 in a call to 3-legged. The response should tell that 3 devices are assigned to the booking but the list of devices would only list 2 of them. Is this what we expect? in that case we need to adapt the descriptions to make this clear.

The 2 APIs are independent and should not be compatible as Booking models are different. I think we don't need to discuss backward compatibility between them. My expectation is that a bookingId from one of the APIs is not expected to work with the other one. Coherently, the endpoint to retrieve bookings for a device is expected to deal with bookings for the same API, not to retrieve bookings from the other.

RamTMO replied:
I have added lots of text discussing the 2-legged and 3-legged sceneraios and use cases.  If a device is assigned via 3-legged access token then it will not show-up in the output list.  I explained it in the description clearly.  Also I have added an enum "DEVICE_DISCLOSURE_ERROR" which the operator can use if the operator system believes identifiable device information SHOULD NOT be disclosed for whatever reason.

RamTMO replied:
Regarding the retrieve: These two APIs are completely independent and the bookingID from one can't be used in another. Agree.  I will keep this endpoint as it is.  In most cases, ony one booking (associated with the current API) will be returned which is in the context of the current API.

jlurien commented on code/API_definitions/qos-booking-and-assignment.yaml:
```diff
@@ -0,0 +1,1644 @@
+        content:
+          application/json:
+            schema:
+              $ref: "#/components/schemas/DeviceAssignmentInput"
```
Same concerns with 2-leeged vs 3-legged

RamTMO replied:
same as above

RamTMO replied:
done

jlurien commented on code/API_definitions/qos-booking-and-assignment.yaml:
```diff
@@ -0,0 +1,1644 @@
+        required: true
+      responses:
+        "201":
+          description: Release of devices successful
```
When providing several devices some may fail

RamTMO replied:
Yes, that is correct.  I have added additional comments.  If some devices fail,  the call will return PARTIAL_SUCCESS with the list of successfully released devices.  The user can also use GET to get the list of still active devices in the booking.

jlurien reviewed (commented):
Thanks for the proposal. As a separate API it makes much more sense to me. My main comments are related to how we deal with 3-legged tokens for this API. We probably need to review the rules we have and adapt them. Copying what we have in other QoD APIs, which deal with just one device. it is not directly applicable here.

RamTMO commented:
@jlurien  Thank you for the great comments.  Will take care of them.  Overall I agree with you regarding 2-legged and 3-legged scenarios.  Murat and I dicussed about it and cleaned-up as much as possible. We will evaluate various use cases and fix the device confidentiality issues towards for the final version.  In the meantime. I have implemented some of the changes related version 0.6 and naming conventions this morning and working on the state related minor changes. 

Also I agree with you regarding 'async' nature of the device assignment. Some thinking is requred to understand whether it make sense or not.  it may be simply a synchronous operation with a binary output.  Since it is not 'required', we can leave it to operator choice -- this one needs discussion along with 2-legged/3-legged issues.

Masa8106 commented:
Thank you for your proposal. 

I think a point of this API is that one Booking ID may associate with multiple devices. The point is not seen in any other QoD related APIs. However, I think we can refer to a relation between Network ID and Device in Dedicated NW Access API. For example, we can consider optional filtering for a given device in GET operation of /devices/{bookingId}, if you are ok.

For code itself, there seems to be some issues as far as I see "[Test wip API](https://github.com/hdamker/ReleaseManagement/issues/24)". If not yet, it would be good if you correct them.

RamTMO commented on code/API_definitions/qos-booking-and-assignment.yaml:
```diff
@@ -0,0 +1,1644 @@
+  url: https://github.com/camaraproject/QoSBooking
+
+servers:
+  - url: "{apiRoot}/qos-booking-and-assignment/vwip"
```
I agree.  Good suggestion. But right now, I keep this still as 'qos-booking-and-assignment'. When all of us ready, we can do it at once.

RamTMO commented on code/API_definitions/qos-booking-and-assignment.yaml:
```diff
@@ -0,0 +1,1644 @@
+        default: http://localhost:9091
+        description: API root, defined by the service provider, e.g. `api.example.com` or `api.example.com/somepath`
+tags:
+  - name: Advance QoS Booking
```
Fixed it in the recent commit - Update3

RamTMO commented on code/API_definitions/qos-booking-and-assignment.yaml:
```diff
@@ -0,0 +1,1644 @@
+tags:
+  - name: Advance QoS Booking
+    description: Book and manage QoS for multiple devices in advance
+  - name: Device QoS Assignment
```
Fixed it in the recent commit - Update3

RamTMO commented on code/API_definitions/qos-booking-and-assignment.yaml:
```diff
@@ -0,0 +1,1644 @@
+    description: Manage the assignment of devices of QoS
+
+paths:
+  /book-qos:
```
It was fixed earlier -- but used /qos-booking. (as singular)

RamTMO replied:
done

RamTMO commented on code/API_definitions/qos-booking-and-assignment.yaml:
```diff
@@ -0,0 +1,1644 @@
+        - Advance QoS Booking
+      summary: Reserve QoS in advance for one or more devices. Devices can be assigned later to this booking
+      description: |
+        Triggers a new booking in advance and assign this reserved booking profile to one or more devices when the devices are ready.
```
Fixed it in the recent commit - Update3

RamTMO commented on code/API_definitions/qos-booking-and-assignment.yaml:
```diff
@@ -0,0 +1,1726 @@
+      default: 86400
+
+    Devices:
+      description: Array of devices to be assigned to the booking and provisioned.  At least one device is required
```
Agree.  I will update the description as such

RamTMO commented on code/API_definitions/qos-booking-and-assignment.yaml:
```diff
@@ -0,0 +1,1726 @@
+        - openId:
+            - "qos-booking-and-assignment:assign-devices-to-booking:create"
+
+  /devices/{bookingId}:
```
I will hold on to this suggestion. Let us discuss first in the call.  I would like to keep it in the devices resource group but it does not matter in theory.

RamTMO replied:
I agree.  For GET operation we can use  GET /devices/{bookingId}/assignments:    it goes along with the same pattern as others in the devices subgroup.  Minor change.  And self explonatory.  Let us discuss

RamTMO replied:
Fixed it as per your suggestion all over.  Now /qos-bookings/ is the only resource group

RamTMO commented:
@jlurien @Masa8106  Incorporated most of the suggestions.  I'm still keeping devices resource group as a separate resouce group. We have two resource groups in this file one is /qos-bookings to manage bookings and the other one is /devices to assign and manage devices. Let us discuss the plus and minuses.   

Also, I have added .feature file with initial set of test definitions.

RamTMO commented:
@Masa8106 @jlurien @hdamker The next set of changes are commits -- all suggestions are incorporated.  Test definition file is in branch and in the same PR.  Please review and let me know if there are any additional sugesstions or changes.

jlurien commented on code/API_definitions/qos-booking-and-assignment.yaml:
```diff
@@ -0,0 +1,1756 @@
+          $ref: "#/components/responses/Generic429"
+      security:
+        - openId:
+            - "qos-booking-and-assignment:retrieve-bookings-of-a-device:read"
```
```suggestion
            - "qos-booking-and-assignment:qos-bookings:retrieve-by-device"
```

hdamker reviewed (changes requested):
Some easy to fix issues

jlurien commented on code/API_definitions/qos-booking-and-assignment.yaml:
```diff
@@ -0,0 +1,1756 @@
+         - For example:  
+           - If there are 10 devices assigned in a booking; but the GET operation returns only 7; It means 3 devices can't be disclosed
+           - The source of truth is in the following properties in the booking details: "totalDevicesInBooking" and "remainingDevicesInBooking".        
+           - In such a situation the operator can set  'statusInfo' as 'DEVICE_DISCLOSURE_ERROR' to clearly communicate the fact that some devices are not disclosed.
```
I wouldn't call it DEVICE_DISCLOSURE_ERROR but DEVICE_DISCLOSURE_POLICY_APPLIED or similar. Instead of "operator can" let's choose one between "operator may/should/must", probably may or should at most.

jlurien commented on code/API_definitions/qos-booking-and-assignment.yaml:
```diff
@@ -0,0 +1,1756 @@
+        The current status of the requests in any of the endpoints in Booking or Assignment. The status can be one of the following:
+        * `PENDING` - The request is stil being processed. For booking, booking is requested and for assignements, the request is pending. 'statusInfo' may have additional information.
+        * `SUCCESSFUL` - The request is processed successfully. This may happen immediately.  The state will go to PENDING and then back to SUCCESSFUL in certain conditions.ƒ
+        * 'PARTIAL_SUCCESS' -  The request is processed with partial success - mostly applicable to device assignments.  'statusInfo' may have additional information. This enum is optional, and useful if the operator wants to communicate partical success.
```
```suggestion
        * `PARTIAL_SUCCESS` -  The request is processed with partial success - applicable to device assignments.  'statusInfo' may have additional information. This enum is optional, and useful if the operator wants to communicate partial success.
```
any case applicable outside device assignments? 

I have doubts around the expectations for the lifecycle of property `status`, which is returned with every BookingInfo and it should represent the status of the Booking. For example you assign several devices and some fail, then status=PARTIAL_SUCCESS, then you add one or many more and all succeed, is status still kept as PARTIAL_SUCCESS or reflects the result of the latest request only. And if a subsequent asssignment fails, is status  moved to FAILURE or still keeps the past PARTIAL_SUCCESS.

jlurien commented on code/API_definitions/qos-booking-and-assignment.yaml:
```diff
@@ -0,0 +1,1756 @@
+        * `DURATION_EXPIRED` - Session terminated due to requested duration expired
+        * `NETWORK_TERMINATED` - Network terminated the QoS Booking for unknown reasons
+        * `DELETE_REQUESTED`- User requested the deletion of the QoS Booking
+        * 'VALIDATION_PENDING' - The validation is pending. The end user will get a callback with the availability of the booking
```
Adjust `` instead of ''

RamTMO replied:
Fixed

jlurien commented on code/API_definitions/qos-booking-and-assignment.yaml:
```diff
@@ -0,0 +1,1756 @@
+
+    StatusSource:
+      description: |
+        Very useful information when a callback is called with full message body
```
Can you please explain the aim of this property? What is the relation with callbacks and what means a "full message body" for a callback. I think we don't have anything similar to this in any other API dealing with callbacks

jlurien commented on code/API_definitions/qos-booking-and-assignment.yaml:
```diff
@@ -0,0 +1,1756 @@
+        * 'VALIDATION_PENDING' - The validation is pending. The end user will get a callback with the availability of the booking
+        * 'SERVICE_NOT_AVAILABLE' - The requested booking could not be fullfilled because of operator could not support the service area/number of devices/qosProfile in the requested time
+        * 'QOS_PROFILE_NOT_SUPPORTED'  - The requested QoS Profile is not supported by the operator
+        * 'BOKING_UNKNOWN_ERROR' - An unknown error happend while booking
```
```suggestion
        * `BOOKING_UNKNOWN_ERROR` - An unknown error happened while booking
```

jlurien commented on code/API_definitions/qos-booking-and-assignment.yaml:
```diff
@@ -0,0 +1,1756 @@
+        * 'QOS_PROFILE_NOT_SUPPORTED'  - The requested QoS Profile is not supported by the operator
+        * 'BOKING_UNKNOWN_ERROR' - An unknown error happend while booking
+        * 'BOOKING_COMPLETED' - The requested booking is commpleted successfully. 
+        * 'BOOKING_INVALID'  -  The given bookingId is invalid, malformed, or not available in the system
```
As bookingId is a path param, this situation should end with a 404 error probably, instead of a 2xx with this value

jlurien replied:
Any error in the backend when reported back to the client it should be 404 so this item is not going to be returned. It may be misleading.

jlurien reviewed (commented):
Thanks for the update, it looks now more mature. I have looked into other aspects of the API and add more comments, mainly:
- Adjust scope naming to the guidelines
- Doubt around the properties dealing with status and expected life cycle of the resources

RamTMO commented on code/API_definitions/qos-booking-and-assignment.yaml:
```diff
@@ -0,0 +1,1756 @@
+          $ref: "#/components/responses/Generic429"
+      security:
+        - openId:
+            - "qos-booking-and-assignment:get-devices-for-a-booking:get"
```
Fixed

RamTMO commented:
@Masa8106 @jlurien @hdamker Fixed .yaml and test file in many places based on your comments.  Please review and let me know if you have any further comments.

jlurien commented on code/API_definitions/qos-booking-and-assignment.yaml:
```diff
@@ -0,0 +1,1741 @@
+    Status:
+      description: |
+        The current status of the requests in any of the endpoints in Booking or Assignment. The status can be one of the following:
+        * `PENDING` - The request is stil being processed. For booking, booking is requested and for assignements, the request is pending. 'statusInfo' may have additional information.
```
```suggestion
        * `PENDING` - The request is still being processed. For booking, booking is requested and for assignments, the request is pending. 'statusInfo' may have additional information.
```

jlurien commented on code/API_definitions/qos-booking-and-assignment.yaml:
```diff
@@ -0,0 +1,1756 @@
+              $ref: "#/components/schemas/DeviceAssignmentInput"
+        required: true
+      responses:
+        "201":
```
I would leave 201 for the "create" operations, just the ones with scope "...:create", and 200 for the rest

jlurien commented on code/API_definitions/qos-booking-and-assignment.yaml:
```diff
@@ -0,0 +1,1741 @@
+          duration: 3600
+          device:
+            phoneNumber: "+123456789"
+          applicationServer:
```
No applicationServer in the model

jlurien commented on code/API_definitions/qos-booking-and-assignment.yaml:
```diff
@@ -0,0 +1,1741 @@
+      summary: List of QoS sessions for the device
+      description: A single QoS session for the device is available
+      value:
+        - bookingId: "3fa85f64-5717-4562-b3fc-2c963f66afa6"
```
Example is not compatible with the schema

RamTMO replied:
Correct.  Removed all examples and bring it back after M2 with a new PR.

jlurien commented on code/API_definitions/qos-booking-and-assignment.yaml:
```diff
@@ -0,0 +1,1741 @@
+        specversion: "1.0"
+        type: "org.camaraproject.qos-booking-and-assignment.v0.status-changed"
+        time: "2024-06-01T13:00:00Z"
+        data:
```
data not compatible with the schema

jlurien commented on code/API_definitions/qos-booking-and-assignment.yaml:
```diff
@@ -0,0 +1,1741 @@
+                code: TOO_MANY_REQUESTS
+                message: Rate limit reached.
+
+  examples:
```
Examples copied from the other API are not compatible with the schemas in this one

RamTMO replied:
I will delete all examples for now  and bring it back after M3 in a separate PR

RamTMO replied:
Deleted all examples for now and bring it back after M3 in a separate PR

jlurien commented on code/API_definitions/qos-booking-and-assignment.yaml:
```diff
@@ -0,0 +1,1741 @@
+      summary: QoS booking status is available
+      description: QoS booking info when status is available
+      value:
+        bookingId: "3fa85f64-5717-4562-b3fc-2c963f66afa6"
```
bookingDetails, etc

RamTMO replied:
Deleted all examples for now and bring it back after M3 in a separate PR

jlurien commented on code/API_definitions/qos-booking-and-assignment.yaml:
```diff
@@ -0,0 +1,1741 @@
+            longitude: 7.10066
+          radius: 100
+
+    BOOKING_UNAVAILABLE:
```
Idem

RamTMO replied:
Deleted all examples for now and bring it back after M3 in a separate PR

jlurien commented on code/API_definitions/qos-booking-and-assignment.yaml:
```diff
@@ -0,0 +1,1741 @@
+      items:
+        $ref: "#/components/schemas/Device"
+
+    ApplicationServer:
```
Not needed indeed, to be removed from the spec

RamTMO replied:
Fixed. Removed ApplicationServer and any unused schemas.

RamTMO commented on code/API_definitions/qos-booking-and-assignment.yaml:
```diff
@@ -0,0 +1,1741 @@
+        * 'SERVICE_NOT_AVAILABLE' - The requested booking could not be fullfilled because of operator could not support the service area/number of devices/qosProfile in the requested time
+        * 'QOS_PROFILE_NOT_SUPPORTED'  - The requested QoS Profile is not supported by the operator
+        * `BOOKING_UNKNOWN_ERROR` - An unknown error happened while booking
+        * 'BOOKING_COMPLETED' - The requested booking is commpleted successfully. 
```
Since devices can be added in batches some earlier, some now, some some later;  We can add two more status enums -- 1. BOOKING_SCHEDULE_STARTED and 2. BOOKING_SCHEDULE_ENDED.   If we were to do this at the device level, it is would be very complicated.

RamTMO commented on code/API_definitions/qos-booking-and-assignment.yaml:
```diff
@@ -0,0 +1,1741 @@
+        - DURATION_EXPIRED
+        - NETWORK_TERMINATED
+        - DELETE_REQUESTED
+        - VALIDATION_PENDING
```
@tlohmar  okay.. got it.  I would use BOOKING_COMPLETED.  Any other suggestion?

tlohmar replied:
I recommend to keep the bookings and the assignments separated. Maybe BOOKING_ACCEPTED or                                                     similar.

RamTMO reviewed (commented):
All are fixed and completed

hdamker commented on code/Test_definitions/qos-booking-and-assignment.feature:
```diff
@@ -0,0 +1,195 @@
+
+Feature: CAMARA QoS Booking and Assignment - v0.1rc1
```
```suggestion
Feature: CAMARA QoS Booking and Assignment, v0.1.0-rc.1
```

hdamker commented on code/API_definitions/qos-booking-and-assignment.yaml:
```diff
@@ -0,0 +1,1563 @@
+        * `ASSIGNMENT_COMPLETED` - The request assignment of devices is successfully completed
+        * `BOOKING_INVALID` - The user requests to release a device from booking, but the booking is invalid
+        * `BOOKING_CANCELLED` - The user requests to release a device from booking, but the booking is cancelled
+        * `BOOKING_EXPIRED` - The user requests to release a device from booking, but the booking is expiered and no more active
```
```suggestion
        * `BOOKING_EXPIRED` - The user requests to release a device from booking, but the booking is expired and no more active
```

hdamker commented on code/API_definitions/qos-booking-and-assignment.yaml:
```diff
@@ -0,0 +1,1563 @@
+        * `DEVICE_NOT_FOUND` - One or some of the devices are not found to validate and associate to booking. 
+        * `DEVICE_UNKNOWN_ERROR` - One, or more, or all of devices are not assigned or not released because of unknown error.
+        * `DEVICE_DISCLOSURE_POLICY_APPLIED` - Some or all of the devices are not disclosed because of privacy, or legal requirements, or for unknown reasons.
+        * `QUOTA_EXCEEDED` - The number of devices requested to assign to a booking is more than the allocated totoal for the booking. Only remaining devices are assigned
```
```suggestion
        * `QUOTA_EXCEEDED` - The number of devices requested to assign to a booking is more than the allocated total for the booking. Only remaining devices are assigned
```

hdamker commented on code/API_definitions/qos-booking-and-assignment.yaml:
```diff
@@ -0,0 +1,1563 @@
+            duration:
+              $ref: "#/components/schemas/Duration"
+            serviceArea:
+              $ref: "#/components/schemas/Area"   
```
Trailing spaces (after M3 we should enable linting in the repository to check both APIs):

```suggestion
              $ref: "#/components/schemas/Area"
```

hdamker commented on code/Test_definitions/qos-booking-and-assignment.feature:
```diff
@@ -0,0 +1,195 @@
+
+
+    # Success scenarios for GET /qos-bookings/{bookingId}
+    # Given a bookingID, this GET operation gets the details of the orignal booking
```
```suggestion
    # Given a bookingID, this GET operation gets the details of the original booking
```

hdamker reviewed (approved):
LGTM, good quality and design for a 0.1.0-rc.1 version.

jlurien reviewed (approved):
I see that the wrong examples are removed, but we should add new meaningful ones. We can do that before M4 along with further refinement in this period, so LGTM for the rc.1

Thanks @RamTMO for the effort in such a short time

jlurien commented:
> LGTM. As this PR is a major change for our repository, I think it would be good for us that all code owners approve it before merging it into main branch.

Yes, we can merge it now and adapt #18 to consider both APIs

RamTMO commented:
Thank you all!  Let me know if I have to anything from my side.
